### ISSUE TITLE
Some questions (HELP 😭 )

### ISSUE THREAD

**Issue by mcziel** [OFFLINE-GmbH/oc-cashier-plugin#23]
Hi,

So, I have a project with a deadline for this next monday 😆  and I haven't integrated stripe yet. I think this plugin accomplishes all I need but I am having some difficulties trying to set it up properly. 

When I try to use the stripeElementsForm component, I am getting this:

<img width="530" alt="Screen Shot 2021-05-26 at 6 50 13 PM" src="https://user-images.githubusercontent.com/6682458/119741197-396deb00-be53-11eb-9e7c-14385bb50672.png">

Am I missing a CSS or something like that?

Also, is just a credit card field enough to get a subscription? I thought I would need something like this:

<img width="410" alt="Screen Shot 2021-05-26 at 6 52 31 PM" src="https://user-images.githubusercontent.com/6682458/119741366-8b167580-be53-11eb-8635-7f4ef20956ac.png">

Would it be possible to get in touch with you through slack, telegram, discord, or any channel where I could bother you a bit with some questions? This is a rush and I'd love to have some quick answers (I swear I won't bother you a lot).

NOTE: I have already set up the raindlab user plugin and all that. Now I need to allow those users to pay a monthly subscription, if they are not subscribed they will be redirected to a payment page.

Please, Help 😭

**Comment by tobias-kuendig** [comment]
Check your JS console, do you get any errors? Also, did you set the `includeStripeJs` and `ìncludeCss` properties to true?

https://github.com/OFFLINE-GmbH/oc-cashier-plugin/blob/master/components/StripeElementsForm.php#L44

**Comment by mcziel** [comment]
Hi,

There's a listener error:

<img width="762" alt="Screen Shot 2021-05-27 at 9 10 17 AM" src="https://user-images.githubusercontent.com/6682458/119831682-82fa1c80-becb-11eb-93a2-308ca49e4718.png">

<img width="539" alt="Screen Shot 2021-05-27 at 9 10 22 AM" src="https://user-images.githubusercontent.com/6682458/119831692-84c3e000-becb-11eb-830b-e613d9107083.png">

And I am setting JS and CSS like this in the plugin component:

<img width="447" alt="Screen Shot 2021-05-27 at 9 07 45 AM" src="https://user-images.githubusercontent.com/6682458/119831256-1848e100-becb-11eb-8013-fa8921937638.png">

**Comment by mcziel** [comment]
The JS error is because I am not passing the cost of the suscription I think, but I am not sure where should I define it.

**Comment by tobias-kuendig** [comment]
Do you have any other plugins installed? The code that generates the error does not look related to this plugin (if you search the repo there is no `createCheckoutSession` function anywhere).

**Comment by mcziel** [comment]
My bad, It was me trying to set up stripe by myself and I forgot to delete the code.

Now I do not have any errors but still getting this:

<img width="295" alt="Screen Shot 2021-05-27 at 9 57 51 AM" src="https://user-images.githubusercontent.com/6682458/119839236-0454ad80-bed2-11eb-911f-a9db4d19cc32.png">

The CSS and JS:

<img width="464" alt="Screen Shot 2021-05-27 at 9 58 34 AM" src="https://user-images.githubusercontent.com/6682458/119839396-22baa900-bed2-11eb-938e-d68dc81a5466.png">

Also I have those options  checked in the  component settings.

**Comment by tobias-kuendig** [comment]
Do you have a `{% scripts %}` placeholder somewhere on your layout? This is required so the JS can be loaded.
https://octobercms.com/docs/markup/tag-scripts

Check your page source if you can see this code (once the placeholder is added to the layout):
https://github.com/OFFLINE-GmbH/oc-cashier-plugin/blob/master/components/stripeelementsform/script.htm#L9

**Comment by mcziel** [comment]
Yes I do:

![Screen Shot 2021-05-27 at 10 35 26 AM](https://user-images.githubusercontent.com/6682458/119845489-50562100-bed7-11eb-8a95-4e8202b223f3.png)

If i check the source this is what I see:

![Uploading Screen Shot 2021-05-27 at 10.37.40 AM.png…]()

**Comment by tobias-kuendig** [comment]
Alright, then start debugging. Add some console.logs or use your brower's script debugger.

Check to see if the `elements` var gets populated (this means Stripe is loaded) and see if the `#card-element` in your DOM is being replaced. You should see a error somewhere or at least be able to find out where the script stops working.

**Comment by mcziel** [comment]
The #card-element has an iframe and It's been loaded:

![Screen Shot 2021-05-27 at 12 14 32 PM](https://user-images.githubusercontent.com/6682458/119861362-4f2bf080-bee5-11eb-8bc8-ac472795f52f.png)

![Screen Shot 2021-05-27 at 12 15 18 PM](https://user-images.githubusercontent.com/6682458/119861381-53f0a480-bee5-11eb-91b5-c4f38d9eeee6.png)

I am not getting an error =/

**Comment by tobias-kuendig** [comment]
So everything seems to be working as expected, but for some reason the iframe is invisible. Please check that no browser extension is removing the frame. Also disable any CSS styles that might interfere with the iframe.

According to your screenshot above everything is working fine

**Comment by mcziel** [comment]
You're right, my css is not working properly with the iframe. I disabled my css and now I can put the credit card, expiration date and zip code, but when I hit the pay with card button, it returns an empty array, I am using a test credit card from https://stripe.com/docs/testing#regulatory-cards

![Screen Shot 2021-05-27 at 6 54 15 PM](https://user-images.githubusercontent.com/6682458/119906820-f4ae8680-bf1c-11eb-9258-ba46c94cdc62.png)

**Comment by mcziel** [comment]
Also I am wondering, how to specify the subscription price? I mean in Stripe Dashboard I set up a product (which is a subscription) but I'm confused about how to set it up with the plugin, also there are some special things the website that needs extra payment, for example, subscribed users can download anything but special songs if they want those special songs they have to single pay for them.

**Comment by tobias-kuendig** [comment]
> Also I am wondering, how to specify the subscription price?

AFAIK you set that on Stripe's end by creating a product. As for how to assign it to a user, check out Laravel Cashier's docs (this plugin is just a wrapper around it):
https://laravel.com/docs/8.x/billing#creating-subscriptions

The code from the docs could be executed in the `offline.cashier::stripeElementForm.submit` event listener as described [in the README](https://github.com/OFFLINE-GmbH/oc-cashier-plugin#handle-form-submission).

> subscribed users can download anything

You can easily check if a user has a subscription by calling `$user->subscribed('your-subscription')`. There is also [the `needsSubscription` componenent](https://github.com/OFFLINE-GmbH/oc-cashier-plugin#needssubscription) to block access to a certain page.

> if they want those special songs they have to single pay for them.

Single charges are supported by Cashier as well: https://laravel.com/docs/8.x/billing#single-charges

**Comment by mcziel** [comment]
Thank you, I am working on what you said.

Why the `ends_at` field is empty? even with a successful subscription?

Is the subscription needed component validating that the payment was successful? Because I created a subscription with pending payment, and I was allowed to navigate to a restricted area.

**Comment by mcziel** [comment]
I think I have to modify this event offline.cashier::subscription.check but I am not pretty sure how to specify that only return true with succesful payments. 

How can I access to the payment error, so I can tell the user what happened?

**Comment by tobias-kuendig** [comment]
> Why the ends_at field is empty? even with a successful subscription?

IIRC this is used if the user has cancelled the subscription. Empty = the subscription is active and not cancelled.


> Is the subscription needed component validating that the payment was successful? Because I created a subscription with pending payment, and I was allowed to navigate to a restricted area.

We simply redirect the call to Cashier's `subscribed` method, so you would have to check there if pending payments are considered as a "valid subscription":

https://github.com/OFFLINE-GmbH/oc-cashier-plugin/blob/master/components/NeedsSubscription.php#L48

> How can I access to the payment error, so I can tell the user what happened?

You can use the webhook events to react to failed payments:
https://github.com/OFFLINE-GmbH/oc-cashier-plugin#handle-stripe-webhooks

For example:

```php
        Event::listen('offline.cashier::stripe.webhook.invoice.payment_failed', function ($payload, $request) {
            $subscription = $payload['data']['object']['subscription'];
            try {
                $user = User::fromSubscriptionId($subscription);
                // Do something with $user
            } catch (\Exception $e) {
                // do nothing.
            }

            logger()->error('Payment failed', compact('user', 'payload'));
        });

```
